curvefs/test/metaserver/s3compactwq_test.cpp

```diff
@@ -0,0 +1,420 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2021-09-07
+ * @Author: majie1
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "curvefs/src/metaserver/s3compact_manager.h"
+#include "curvefs/src/metaserver/s3compact_wq_impl.h"
+#include "curvefs/test/metaserver/mock_s3_adapter.h"
+#include "curvefs/test/metaserver/mock_s3_s3compact_impl.h"
+
+using ::testing::_;
+using ::testing::AtLeast;
+using ::testing::DoAll;
+using ::testing::Return;
+using ::testing::ReturnArg;
+using ::testing::SaveArg;
+using ::testing::SetArgPointee;
+using ::testing::StrEq;
+
+namespace curvefs {
+namespace metaserver {
+class S3CompactWorkQueueImplTest : public ::testing::Test {
+ public:
+    void SetUp() override {
+        opts_.enable = true;
+        opts_.threadNum = 1;
+        opts_.queueSize = 1;
+        opts_.blockSize = 4;
+        opts_.chunkSize = 64;
+        opts_.fragmentThreshold = 20;
+        opts_.maxChunksPerCompact = 10;
+        s3adapter_ = std::make_shared<MockS3Adapter>();
+        inodeStorage_ = std::make_shared<MemoryInodeStorage>();
+        impl_ = std::make_shared<S3CompactWorkQueueImpl>(s3adapter_, opts_);
+        mockImpl_ =
+            std::make_shared<MockS3CompactWorkQueueImpl>(s3adapter_, opts_);
+    }
+
+    void TearDown() override {
+        return;
+    }
+
+ protected:
+    S3CompactWorkQueueOption opts_;
+    std::shared_ptr<MockS3Adapter> s3adapter_;
+    std::shared_ptr<InodeStorage> inodeStorage_;
+    std::shared_ptr<S3CompactWorkQueueImpl> impl_;
+    std::shared_ptr<MockS3CompactWorkQueueImpl> mockImpl_;
+};
+
+TEST_F(S3CompactWorkQueueImplTest, test_GetNeedCompact) {
+    // no need compact
+    ::google::protobuf::Map<uint64_t, S3ChunkInfoList> s3chunkinfoMap;
+    S3ChunkInfoList l1;
+    for (int i = 0; i < 10; i++) {
+        auto ref = l1.add_s3chunks();
+        ref->set_chunkid(i);
+    }
+    s3chunkinfoMap.insert({1, l1});
+    ASSERT_TRUE(impl_->GetNeedCompact(s3chunkinfoMap).empty());
+
+    // need compact
+    S3ChunkInfoList l2;
+    for (int i = 0; i < 30; i++) {
+        auto ref = l2.add_s3chunks();
+        ref->set_chunkid(i);
+    }
+    s3chunkinfoMap.insert({2, l2});
+    ASSERT_EQ(impl_->GetNeedCompact(s3chunkinfoMap).size(), 1);
+
+    // too much need compact, control size
+    for (int j = 3; j < 20; j++) {
+        S3ChunkInfoList l;
+        for (int i = 0; i < 30; i++) {
+            auto ref = l.add_s3chunks();
+            ref->set_chunkid(i);
+        }
+        s3chunkinfoMap.insert({j, l});
+    }
+    ASSERT_EQ(impl_->GetNeedCompact(s3chunkinfoMap).size(), 10);
```

Comment:
assert with opts_.maxChunksPerCompact instead of 10

Reply:
fix